extract kindsys to external library

mildwonkey opened the pull request:
This PR removes `pkg/kindsys` in favor of https://github.com/grafana/kindsys, which is _mostly_ a direct port of said package. some of the loading functions that had been in kindsys are now under `pkg/cuectx`; the majority of the remaining changes are matters of renaming things.

The most interesting difference is how the .cue files embedded in `kindsys` are now brought into grafana for generating kinds:

https://github.com/grafana/grafana/compare/mildwonkey/extract-kindsys?expand=1#diff-75c09abd49577c168b0056d3bebe275d54081c4b5a88879e0f47eaf97ae06b41R129-R131

and the method used in the plugindef generation:

https://github.com/grafana/grafana/compare/mildwonkey/extract-kindsys?expand=1#diff-75c09abd49577c168b0056d3bebe275d54081c4b5a88879e0f47eaf97ae06b41R198-R203

"Was that actually interesting" is a valid question right about now. It might not be _interesting_, but it turned out to be relatively simple to start from whatever root repository and merge in the kindsys cue files as dependencies; now that I've demonstrated a general pattern for passing around a filesystem of cue files we can do something similar if/when it's time to move other cue files elsewhere.

This is just step one; some of the plugin related code (`pfs`) may move out of grafana eventually, and I'd love to see all the grafana kinds and generators move into the SDK or another location which can be imported by other callers.

mildwonkey committed:
extract kindsys

mildwonkey committed:
tht one, too

mildwonkey committed:
add kindsys filesystem

mildwonkey committed:
freshly generated kinds

mildwonkey committed:
revert local dev gomod replacements

mildwonkey committed:
go mod tidy

mildwonkey committed:
remove kindsys codeowners

sdboyer commented:
same link twice in the description - mis-paste?

mildwonkey commented:
NOTE WELL: this is going to break the grafana-app-sdk; I need to open a PR over there as well before merging this change.

sdboyer commented on pkg/kindsys/report.go:
```diff
@@ -1,390 +0,0 @@
-//go:build ignore
```
need to preserve this somewhere, and update what's in the Makefile to point to its new location.

wherever the `report.json` pops out, we'll also need to update the kinds dashboard on ops to point to the new location

sdboyer reviewed (changes requested):
overall this seems really close - the only issue i see is preserving the `report.go` which is currently in `pkg/kindsys`. It depends on the generated core registries, so gotta keep it in grafana/grafana.

`pkg/cuectx` seems a little weird, but i don't have a strong opinion on where it goes.

suntala commented on pkg/cuectx/load.go:
```diff
@@ -0,0 +1,226 @@
+// ["cuelang.org/go/cue/load".Config.Package]. If the CUE package to be loaded
+// is the same as the parent directory name, it should be omitted.
+//
+// NOTE this function will be deprecated in favor of a more generic loader
```
Will the more generic loader remain in pkg/cuectx? I am more generally wondering about how to decide what should go to `kindsys` vs `pkg/cuectx`. I'm asking out of curiosity because I'm trying to learn more about this topic.

mildwonkey replied:
Good question - I plan on adding a generic loader function in kindsys and use that to replace the grafana-specific loader (the generic loader will be also be used in the SDK).

mildwonkey committed:
reinstate kindsys report

This may end up living somewhere else (or not! who knows!), but the important part is that I don't get rid of it right now :)

I hate the package layout (kindsysreport/codegen) for the main function and will take pretty much any alternative suggestion, but we can change also change it later.

Note that the generated report.json is in a different location - anything using this (ops something) needs to be updated.

mildwonkey commented:
@sdboyer : oh heck thank you! I reinstated the report.json - the file location is different, but I have absolutely 0 attachment to my choice of filenames & paths. I'd be just as happy putting that code back exactly where it was under `pkg/kindsys/` or pretty much anywhere else.  I'm not trying hard to come up with the bestest location is because I expect the entire report will end up living elsewhere; I imagine most of the generated kinds & related code will end up in the SDK. (but it might not! but also it's really easy to move as often as we need! the stakes are VERY LOW! 😂)

mildwonkey committed:
go mod updates

mildwonkey committed:
kindsysreport in codeowners

sdboyer reviewed (approved):
lovely. LGTM!
